api/src/org/labkey/api/util/SessionAppender.java

```diff
@@ -54,7 +69,7 @@ protected SessionAppender(String name, Filter filter, Layout<? extends Serializa
 
         final String key;
         boolean on;
-        final List<LogEvent> list = new LinkedList<>();
+        final Map<LogEvent, String> eventIdMap = new WeakHashMap<>();
```

Comment:
I think this should probably be a ConcurrentReferenceHashMap.  These can be accessed from multiple threads, so that will at least make the map operations thread safe.  See appenderInfos implementation of that map type

Reply:
Possibly something like this?  Still not great to just randomly remove index(0) but not sure how to get the eldest entry without using a LinkedHashMap, which does not have a weak reference option.

```suggestion
        final Map<LogEvent, String> eventIdMap = new ConcurrentReferenceHashMap<>(16, ConcurrentReferenceHashMap.ReferenceType.WEAK)
        {
            @Override
            public String put(LogEvent key, String value)
            {
                // Safeguard against runaway size. ConcurrentReferenceHashMap does not have removeEldestEntry so have  
                // to randomly remove an entry
                if (size() > 1000)
                    remove(0);

                return super.put(key, value);
            }
        };
```